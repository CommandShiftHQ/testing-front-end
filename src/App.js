import React, { Component } from "react";
import "./styles/App.css";
import ArtistsList from "./components/artists-list";
import FormSection from "./components/form-section";
import Button from "./components/buttons";
import { exportAllDeclaration } from "@babel/types";

const validate = str => {
  if(str.length < 1 || undefined || null) {
    return false;
  }

  return true;
}

class App extends Component {
  constructor() {
    super();
    this.state = {
      artists: [],
      showInput: false,
      validationErrorMessage: ''
    };
  }

  componentDidMount() {
    fetch("https://music-library-api.herokuapp.com/artists")
      .then(response => response.json())
      .then(artists => this.setState({ artists }));
  }

  handleInputSwitch = () => this.setState({ showInput: !this.state.showInput });

  handleOnSubmit = event => {
    event.preventDefault();

    const name = event.target.name.value;
    const genre = event.target.genre.value;

    if(validate(name) && validate(genre)) {

    const options = {
      method: "POST",
      body: JSON.stringify({
        name: name,
        genre: genre
      }),
      headers: {
        "Content-Type": "application/json"
      }
    };

    fetch("https://music-library-api.herokuapp.com/artists", options)
      .then(response => response.json())
      .then(parsedResponse => {
        const newArtist = [
          {
            name: name,
            genre: genre,
            _id: parsedResponse._id
          }
        ];

        this.setState({
          artists: this.state.artists.concat(newArtist),
          showInput: false
        });
      });
    }

    this.setState({
      validationErrorMessage: "Please watch out"
    })
  };

  handleDelete = (event, artistId) => {
    const options = {
      method: "DELETE"
    };

    fetch(
      `https://music-library-api.herokuapp.com/artists/${artistId}`,
      options
    ).then(response => {
      this.setState({
        artists: this.state.artists.filter(artist => artist._id !== artistId)
      });
    });
  };

  render() {
    const { artists, showInput, validationErrorMessage } = this.state;

    if (artists.length === 0) return <div>Loading...</div>;

    return (
      <div className="App">
        <h1>My favourite artists</h1>
        <ArtistsList 
          artists={artists} 
          handleOnClick={this.handleDelete} />
        <div className="artists-input-section">
          {!showInput &&
            <Button
              qaClass="button-add-more"
              label="Add more"
              type="button"
              handleOnClick={this.handleInputSwitch}
            />}
          <FormSection 
            showInput={showInput} 
            validationErrorMessage={validationErrorMessage}
            handleInputSwitch={this.handleInputSwitch} 
            handleOnSubmit={this.handleOnSubmit}/>
        </div>
      </div>
    );
  }
}

export default App;
