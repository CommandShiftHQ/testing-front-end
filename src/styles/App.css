.App {
  text-align: center;
}

.artists-list-item {
  list-style: none;
}

.artists-list-button {
  margin-left: 15px;
  background-color: #f44336;
  color: white;
}
