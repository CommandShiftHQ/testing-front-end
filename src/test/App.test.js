import React from "react";
import ReactDOM from "react-dom";
import { configure, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import App from "../App";

configure({ adapter: new Adapter() });

const mockFetch = data => {
  return jest.fn().mockImplementation(() =>
    Promise.resolve({
      ok: true,
      json: () => data
    })
  );
};

const artists = [
  { name: "Tame Impala", genre: "Rock", _id: 112233 },
  { name: "Radiohead", genre: "Alternative", _id: 223344 },
];

test("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test("the component renders correctly upon initialising", () => {
  const renderedComponent = mount(<App />);
  expect(renderedComponent).toMatchSnapshot();
});

test("fetches data and sets it in state when the component mounts", async () => {
  global.fetch = mockFetch(artists);

  const renderedComponent = await mount(<App />);

  await renderedComponent.update();

  expect(global.fetch).toHaveBeenCalled();
  expect(global.fetch).toHaveBeenCalledWith("https://music-library-api.herokuapp.com/artists");
  expect(renderedComponent.state("artists").length).toEqual(2);
});

test("if no data should display a loading div", () => {
  const renderedComponent =  mount(<App />);
  const loadingDiv = renderedComponent.find('div');

  expect(loadingDiv.text()).toEqual('Loading...');
  expect(loadingDiv.html()).toEqual('<div>Loading...</div>');
})

test('displays the items in state as a list and a add more button', () => {
  const renderedComponent = mount(<App />);

  renderedComponent.setState({artists});
  renderedComponent.update();

  expect(renderedComponent.find('li').length).toEqual(2)
  expect(renderedComponent.find('.button-add-more').text()).toEqual('Add more')
})

test('toggles between button and form', () => {  
  const renderedComponent = mount(<App />);

  renderedComponent.setState({artists});
  renderedComponent.update();

  const addMoreButton = renderedComponent.find('.button-add-more');

  addMoreButton.simulate('click');

  expect(renderedComponent.state('showInput')).toEqual(true);
  expect(renderedComponent.find('form').children().length).toEqual(4);
  expect(renderedComponent.find('.button-add-more')).toEqual({});
})