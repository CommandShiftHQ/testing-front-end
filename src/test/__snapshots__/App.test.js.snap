// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the component renders correctly upon initialising 1`] = `ReactWrapper {}`;
